### PR TITLE
LUCENE-9823: Fix not to rewrite boosted single term SynonymQuery

### PR DESCRIPTION
**Pull request by chlorochrule**
# Description

> When rewriting a SynonymQuery with a single term, we create a BoostQuery wrapping a TermQuery. 

However, 

>  it now multiplies the final TermQuery score instead of multiplying the term frequency before it's passed to the score calculation.

JIRA: https://issues.apache.org/jira/browse/LUCENE-9823

# Solution

Remove unsafe codes of `SynonymQuery` and `CombinedFieldQuery`.

# Tests

Added:
- `TestSynonymQuery#testRewrite`

Fixed:
- `TestCombinedFieldQuery#testRewrite`

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/lucene/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] ~I have created a Jira issue and added the issue ID to my pull request title.~ **I've picked newdev ticket**
- [x] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.

### CODE REVIEW

**Review by jtibshirani — COMMENTED** [review]
Thanks @chlorochrule for fixing this. A couple comments:
* Although it's a simple fix, it's always good if changes are accompanied by a unit test. We could add a short test in `TestSynonymQuery` covering rewrites, similar to existing tests we have like `TestBoostQuery#testRewrite`.
* Could we include a similar fix in `CombinedFieldQuery#rewrite`, which has the same problem?

**Comment by chlorochrule** [comment]
Thanks for reviewing, @jtibshirani !
I added `TestSynonymQuery#testRewrite` and fixed the same problem of `CombinedFieldQuery`. I may not fully understand the meaning of:
> The same consideration affects CombinedFieldQuery in sandbox.

If the fix of bc85a2a is  incorrect, please explain what is the same consideration. 

Thanks.

File: lucene/core/src/test/org/apache/lucene/search/TestSynonymQuery.java [hunk 1]
```diff
@@ -466,4 +467,26 @@ public void testRandomTopDocs() throws IOException {
     reader.close();
     dir.close();
   }
+
+  public void testRewrite() throws IOException {
+    // zero length SynonymQuery is rewritten
+    SynonymQuery q = new SynonymQuery.Builder("f").build();
```
Comment (jtibshirani) [inline lucene/core/src/test/org/apache/lucene/search/TestSynonymQuery.java:473]: A small comment: in most other rewrite tests (like `TestBoostQuery#testRewrite`) we check the higher-level call `IndexSearcher#rewrite`. It'd be nice to do this to be consistent and to exercise the full rewrite logic.

**Review by jtibshirani — COMMENTED** [review]
It's looking very close, once these comments are addressed I'll run tests + merge.

**Comment by chlorochrule** [comment]
Thanks for explaining and reviewing again!
I fixed :)

**Review by jtibshirani — APPROVED** [review]
This looks good to me. I also pushed an update to `CHANGES.txt`.

**Pull request merged**